cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -369,63 +384,55 @@ class TypedConverter : public SeqConverter {
 template <typename BuilderType, class Derived>
 class TypedConverterVisitor : public TypedConverter<BuilderType> {
  public:
-  Status AppendData(PyObject* obj, int64_t size) override {
+  Status AppendSingle(PyObject* obj) override {
+    // static_cast<> below devirtualizes the method calls (for performance?)
+    if (obj == Py_None) {
+      return static_cast<Derived*>(this)->AppendNull();
+    } else {
+      return static_cast<Derived*>(this)->AppendItem(obj);
+    }
+  }
+
+  Status AppendMultiple(PyObject* obj, int64_t size) override {
     /// Ensure we've allocated enough space
     RETURN_NOT_OK(this->typed_builder_->Reserve(size));
     // Iterate over the items adding each one
     if (PySequence_Check(obj)) {
       for (int64_t i = 0; i < size; ++i) {
         OwnedRef ref(PySequence_GetItem(obj, i));
-        if (ref.obj() == Py_None) {
-          RETURN_NOT_OK(this->typed_builder_->AppendNull());
-        } else {
-          RETURN_NOT_OK(static_cast<Derived*>(this)->AppendItem(ref));
-        }
-      }
-    } else if (PyObject_HasAttrString(obj, "__iter__")) {
-      PyObject* iter = PyObject_GetIter(obj);
-      OwnedRef iter_ref(iter);
-      PyObject* item;
-      int64_t i = 0;
-      // To allow people with long generators to only convert a subset, stop
-      // consuming at size.
-      while ((item = PyIter_Next(iter)) && i < size) {
-        OwnedRef ref(item);
-        if (ref.obj() == Py_None) {
-          RETURN_NOT_OK(this->typed_builder_->AppendNull());
-        } else {
-          RETURN_NOT_OK(static_cast<Derived*>(this)->AppendItem(ref));
-        }
-        ++i;
-      }
-      if (size != i) {
-        RETURN_NOT_OK(this->typed_builder_->Resize(i));
+        RETURN_NOT_OK(static_cast<Derived*>(this)->AppendSingle(ref.obj()));
       }
     } else {
-      return Status::TypeError("Object is not a sequence or iterable");
+      return Status::TypeError("Object is not a sequence");
     }
     return Status::OK();
   }
+
+ protected:
+  // Append a non-missing item
+  virtual Status AppendItem(PyObject* obj) = 0;
```

Comment:
Making this virtual has performance implications because of inlining (this module is using CRTP to simulate virtual functions without the dynamic dispatch); in practice it may not matter, though

Reply:
Actually this PR is not ondoing the CRTP trick, so I don't think it has performance implications?

I was adding this for two reasons:
1) for consistency with `AppendNull` which has a default version defined on `TypedConverterVisitor` (it only needs to be specialized for struct types)
2) for documentation, since it makes it more obvious that subclasses are supposed to define `AppendItem`

Should I remove it? Or is there a way to have a non-virtual unimplemented base method?

Reply:
This is OK as is; when we can demonstrate real performance benefits to using non-virtual calls we can do some refactoring later

Reply:
Hmm, actually you may be right that even the `static_cast<>` won't be enough to devirtualize the method call (since we may be given a derived type defined somewhere).

I ran a simple micro-benchmark:
```
$ python -m timeit -s "import pyarrow as pa; ty=pa.int64(); data=[42]*1000000" "pa.array(data, type=ty)"
```
* before: 10 loops, best of 3: 42.4 msec per loop
* after: 10 loops, best of 3: 45.5 msec per loop

So there is a slight slowdown indeed...

Reply:
Ok, I undid this. We're now down to 43.3 msec per loop in the micro-benchmark above (so a tiny bit slower than master, but presumably good enough).